### PR TITLE
Normalized metadata, updated metadata pattern

### PR DESCRIPTION
**Pull request by Victor-C-Zhang**
Implemented new metadata pattern with bits for free/last segment

Have metadata interspersed with data instead in a header chunk
- Increases length of continuous allocations to 4095
- IMPORTANT: Allows allocations to span multiple pages without colliding with metadata header chunks
- Slight performance hit bc of misalignment with cache lines and larger data types

### CODE REVIEW

File: asst2/CODE/direct_mapping.c [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+void dm_write_metadata(metadata* curr, size_t size, int occupied, int last) {
+  curr -> size = ((uint8_t)(size-1)) | (occupied << 7) | (last << 6);
+}
+
+void dm_allocate_block(metadata* curr, size_t size) {
+  uint8_t curr_seg = dm_block_size(curr);
+  int is_last_segment = dm_is_last_segment(curr);
+  if (size < curr_seg) {
+    dm_write_metadata(curr, size, 1, 0);
+    dm_write_metadata(curr+size*SEGMENTSIZE, curr_seg-size, 0, is_last_segment);
```
Comment (ihartb) [inline asst2/CODE/direct_mapping.c:25]: would the free write be:
`dm_write_metadata( (curr+1)+size*SEGMENTSIZE, curr_seg-size-sizeof(metadata), 0, is_last_segment);
`
well i guess the free write can't account for the byte of metadata bc our size is in terms of segments, so i guess how do we account for it / do we not need to and why ? 
tl;dr how do we account for the one byte of metadata space we add bc of splitting of the segments

File: asst2/CODE/my_malloc.c [hunk 2]
```diff
@@ -86,8 +65,8 @@ void printMemory() {
 		int j = 0;
 		while (j < num_segments) {
 			metadata* mdata = start + j;
-			char curr_seg = block_size(mdata);
-			int isOcc = is_occupied(mdata);
+			char curr_seg = dm_block_size(mdata);
+			int isOcc = dm_block_occupied(mdata);
 			printf("page[%d][%d] pid %d %p: %d  %hu %p\n", i, j, pdata->pid, mdata, isOcc, curr_seg, (start+num_segments) + j*SEGMENTSIZE);
 			j+=curr_seg;
```
Comment (ihartb) [inline asst2/CODE/my_malloc.c:73]: `j = j+ sizeof(metadata)+curr_seg*SEGMENTSIZE`

File: asst2/CODE/my_malloc.c [hunk 3]
```diff
@@ -127,14 +107,13 @@ void* find_free_page(size_t size, uint32_t curr_id) {
...
+				unsigned char curr_seg = dm_block_size(curr);
+				if (!dm_block_occupied(curr) && (segments_alloc <= curr_seg ) ) {
+					dm_allocate_block(curr, segments_alloc);
 					//segment memory space = start + num_segments
 					//free segment = segment memory space + seg_index * SEGMENTSIZE;
-					void* ret = (void*) ( (start+num_segments) + seg_index*SEGMENTSIZE);
-					return ret;
+					return (void*) (curr + 1);
 				}
 				seg_index+=curr_seg;
```
Comment (ihartb) [inline asst2/CODE/my_malloc.c:120]: `seg_index = seg_index+sizeof(metadata)+curr_seg*SEGMENTSIZE`

File: asst2/CODE/my_malloc.c [hunk 4]
```diff
@@ -86,8 +65,8 @@ void printMemory() {
 		int j = 0;
 		while (j < num_segments) {
```
Comment (ihartb) [inline asst2/CODE/my_malloc.c:66]: `j < page_size`

File: asst2/CODE/my_malloc.c [hunk 5]
```diff
@@ -127,14 +107,13 @@ void* find_free_page(size_t size, uint32_t curr_id) {
 			metadata* start = (metadata*) (mem_space + index*page_size);
 			int seg_index = 0;
 			while ( seg_index < num_segments) {
```
Comment (ihartb) [inline asst2/CODE/my_malloc.c:111]: `seg_index < page_size`

File: asst2/CODE/my_malloc.c [hunk 6]
```diff
@@ -127,14 +107,13 @@ void* find_free_page(size_t size, uint32_t curr_id) {
...
-				metadata* curr = start + seg_index;
-				unsigned char curr_seg = block_size(curr);
-				if (!is_occupied(curr) && (segments_alloc <= curr_seg ) ) { 
-					write_occupied_size(curr, curr_seg, segments_alloc);
+				metadata* curr = start + seg_index*SEGMENTSIZE;
```
Comment (ihartb) [inline asst2/CODE/my_malloc.c:112]: `metadata* curr = start + seg_index;`

File: asst2/CODE/my_malloc.c [hunk 7]
```diff
@@ -201,45 +182,48 @@ int free_ptr(void* p) {
 	int seg_index = (firstMalloc == 1) ? num_segments: 0;
 	//find the pointer to be free and keep track of the previous block incase it is free so we can combine them and avoid memory fragmentation
 	while (seg_index < num_segments) {
```
Comment (ihartb) [inline asst2/CODE/my_malloc.c:186]: `seg_index < page_size`

File: asst2/CODE/my_malloc.c [hunk 8]
```diff
@@ -201,45 +182,48 @@ int free_ptr(void* p) {
 	int seg_index = (firstMalloc == 1) ? num_segments: 0;
 	//find the pointer to be free and keep track of the previous block incase it is free so we can combine them and avoid memory fragmentation
 	while (seg_index < num_segments) { 
-		metadata* curr = start + seg_index;
-		unsigned char curr_seg = block_size(curr);
+		metadata* curr = start + seg_index*SEGMENTSIZE;
```
Comment (ihartb) [inline asst2/CODE/my_malloc.c:187]: `metadata* curr = start + seg_index;`

File: asst2/CODE/my_malloc.c [hunk 9]
```diff
@@ -201,45 +182,48 @@ int free_ptr(void* p) {
...
-				return 1; 
+			if ( !dm_block_occupied(curr) ) {
+				return 1;
 			}
 			else {
-				write_free_size(curr, curr_seg); //free the current pointer
+				// free current pointer
+				dm_write_metadata(curr, curr_seg, 0, dm_is_last_segment(curr));
 				//if the next block is within segment and is free, combine if with my current pointer as a free block
 				if (seg_index+curr_seg < num_segments){
```
Comment (ihartb) [inline asst2/CODE/my_malloc.c:199]: `if (seg_index+sizeof(metadata)+curr_seg*SEGMENTSIZE < page_size)`

File: asst2/CODE/my_malloc.c [hunk 10]
```diff
@@ -201,45 +182,48 @@ int free_ptr(void* p) {
...
-					metadata* next = start + seg_index + curr_seg;
-					if (!is_occupied(next)){
-						write_free_size(curr, curr_seg+block_size(next));
+					metadata* next = curr + curr_seg*SEGMENTSIZE;
```
Comment (ihartb) [inline asst2/CODE/my_malloc.c:200]: `metadata* next = curr+sizeof(metadata)+curr_seg*SEGMENTSIZE`

File: asst2/CODE/my_malloc.c [hunk 11]
```diff
@@ -201,45 +182,48 @@ int free_ptr(void* p) {
...
+					if (!dm_block_occupied(next)){
+						dm_write_metadata(curr, curr_seg + dm_block_size(next), 0,
```
Comment (ihartb) [inline asst2/CODE/my_malloc.c:202]: i am not sure but im guessing we don't have to set the size to be `curr_seg + dm_block_size(next)+sizeof(metadata)` bc the size is in segments, not bytes?

File: asst2/CODE/my_malloc.c [hunk 12]
```diff
@@ -201,45 +182,48 @@ int free_ptr(void* p) {
...
+												dm_is_last_segment(next));
 					}
 				}
 				//if previous block was free, combine my current pointer with previous block
-				if (prevFree) write_free_size(prev, prevSize+block_size(curr));
-				if ( !is_occupied(start) && block_size(start) == num_segments) {
+				if (prevFree) dm_write_metadata(prev, prevSize + dm_block_size(curr),
```
Comment (ihartb) [inline asst2/CODE/my_malloc.c:207]: again, do we have to account for the one metadata space is also free'd out when free'ing the prev block?

File: asst2/CODE/my_malloc.c [hunk 13]
```diff
@@ -201,45 +182,48 @@ int free_ptr(void* p) {
...
+																		0, dm_is_last_segment(curr));
+				if (!dm_block_occupied(start) && dm_block_size(start) == num_segments) {
 					write_free_page( (pagedata*)myblock + page_index, page_index );
 				}
 			}		
-  			exit_scheduler(&timer_pause_dump);
+			exit_scheduler(&timer_pause_dump);
```
Comment (ihartb) [inline asst2/CODE/my_malloc.c]: don't need `exit_scheduler` here, only in `mydeallocation`

File: asst2/CODE/my_malloc.c [hunk 14]
```diff
@@ -201,45 +182,48 @@ int free_ptr(void* p) {
...
-		if (!is_occupied(curr)) {
+		if (!dm_block_occupied(curr)) {
 			prevFree = 1;
 			prevSize = curr_seg;
 		}
 		else { 
 			prevFree = 0;
 		}
-		seg_index= seg_index + curr_seg;
+		seg_index += curr_seg;
```
Comment (ihartb) [inline asst2/CODE/my_malloc.c:223]: `seg_index = seg_index + sizeof(metadata)+ curr_seg*SEGMENTSIZE` ?

**Review by ihartb — COMMENTED** [review]
hey, i guess i'm having issues understanding whether we need to account for the `sizeof(metadata)` when traversing `myblock` and setting `malloc`'d/`free`'d size. its seems all rosy until we get to free'ing where it gets janky for me. eg, if we are combining multiple free blocks where metadata spaces are also being free'd. in this case, how do we set free'd size to include the newly free'd metadata if our size value is in terms of segments while our metadata is only a byte? i guess is the normalization the way we account for this ?

**Comment by Victor-C-Zhang** [comment]
> hey, i guess i'm having issues understanding whether we need to account for the `sizeof(metadata)` when traversing `myblock` and setting `malloc`'d/`free`'d size. its seems all rosy until we get to free'ing where it gets janky for me. eg, if we are combining multiple free blocks where metadata spaces are also being free'd. in this case, how do we set free'd size to include the newly free'd metadata if our size value is in terms of segments while our metadata is only a byte? i guess is the normalization the way we account for this ?

everything in the metadata is in terms of segments, not individual bytes. every time we allocate, we convert the size of allocation into a number of segments to allocate. I’ll double-check that we’re doing this properly, but the size of allocation that the user requests is irrelevant to us past this first step where we decide number of segments.

File: asst2/CODE/direct_mapping.c (hunk 1 above)
Comment (Victor-C-Zhang) [reply asst2/CODE/direct_mapping.c:25]: we don’t account for size of metadata in our program. we get handed a number of segments to allocate and we do that. I guess it’s implied that we’ll need to care about the 1 metadata byte upon allocation

File: asst2/CODE/my_malloc.c [hunk 15]
```diff
@@ -86,8 +65,8 @@ void printMemory() {
 		int j = 0;
 		while (j < num_segments) {
 			metadata* mdata = start + j;
```
Comment (ihartb) [inline asst2/CODE/my_malloc.c]: `metadata* mdata = start + j*SEGMENTSIZE;`

**Review by ihartb — COMMENTED** [review]
erase brain fart
